### PR TITLE
mrc-697: upgrade hintr and the whole app

### PR DESCRIPTION
**Pull request by richfitz**
This PR will allow fairly graceful upgrades of hintr, or of all containers.

For hintr upgrade we

1. pull a new hintr image
2. tell workers to stop once they've finished any jobs
3. stop the hintr server
4. bring up the new hintr server
5. bring up new workers

This will leave current sessions unaffected, and current running jobs will continue.  There will be a period of hintr outage, hopefully < 3 s

For the full upgrade we are less graceful and simply

1. pull all images
2. stop all containers (gracefully, slower than kill)
3. start the whole thing again

This will close all client connections and require re-login.

I am not sure if we should call this "upgrade" or "update". Thoughts?

This PR depends on #14 and https://github.com/reside-ic/constellation/pull/5 and the `.travis.yml` should be updated after the constellation PR is merged.

### CODE REVIEW

**Comment by richfitz** [comment]
To address [this comment](https://github.com/reside-ic/constellation/pull/5#issuecomment-557082557) on the corresponding constellation PR, I have had a go at finding a cleaner separation in the logic between the constellation and hint-deploy bits. Now only the "upgrade hintr" bits are implemented in hint-deploy with the full upgrade path done in constellation. I think this leaves a tidier bit of logic in the `main()` function, with `hint_user` and `hint_upgrade_hintr` switching out early to represent logic specific to this app and everything else being shareable with other constellations (montagu, orderly-web)

**Comment by richfitz** [comment]
(also reminder that we should update the .travis.yml before merging)

File: src/hint_deploy.py [hunk 1]
```diff
@@ -126,6 +129,25 @@ def hint_constellation(cfg):
...
+    docker_util.image_pull(hintr.name, str(obj.data.hintr_ref))
+    docker_util.image_pull(hintr.name, str(obj.data.hintr_worker_ref))
+
+    if container:
+        if container.status == "running":
+            print("Stopping previous hintr and workers")
+            container.exec_run(["hintr_stop"])
+        docker_util.container_remove_wait(container)
+
+    obj.start(subset=[hintr.name, worker.name])
```
Comment (hillalex) [inline src/hint_deploy.py]: So, to be clear, we're not removing the existing workers at all, just adding more new ones? But the hintr stop command prevents the outdated ones from accepting jobs? Is there some cleanup mechanism for the old workers?

Comment (richfitz) [reply src/hint_deploy.py]: The old workers will exit when they finish their last job, but will kick around in exited state. Docker has a server-side AutoRemove option which would help, but that has caused race conditions elsewhere and so I've not used it. We should arrange to periodically run `docker container prune` on the server which will clean out the old exited containers

**Pull request merged**